### DISCUSSION TITLE
Song randomizer in the "now playing" field not working as excepted.

### DISCUSSION THREAD

**Discussion by Jacie1krece** [ppy/osu#18139] [Q&A] [answered]
Hi,
when checking osu!lazer for a behavior that existed before (something simlar to the first-born syndrome, but with songs automatically playing (leaving osu! on the main menu for it to play randomly)), it somehow was bugged/changed. First, all of the "date added" sorting was completly random for seemingly no reason (For future reference, the bottom-most songs were No Title, Happppy Song, iiillluuuvvvUUU, Happppy Song and Raven Haven - it has been that way in osu!lazer since I first imported songs from osu!stable), and the old "date added" sorting is used as a reference in the automatic playing - newest additions being suspiciously together with SOOOO's song confirming that behavior. Could someone tell me please, what happened and is it a bug?

**Comment by bdach** [comment]
i'm not sure what this is talking about but i believe this may be a duplicate of https://github.com/ppy/osu/issues/13717

**Reply by Jacie1krece** [reply]
Yes, that issue is (or was) happening, but it isn't the problem (It only made that somehow all the SOOO songs were together at the beginning). To be clearer, my issue is about the "song randomiser" used for playing songs on the main menu randomly isn't random, but instead using the "date added" sort (or at least, the old one with the issue you provided, because the new one is different).

**Comment by peppy** [comment, marked as answer]
As for fixing the randomisation, we likely need to better define what we want.

Currently:
- `MusicController` has a stable `next`/`prev` operation that offers no randomisation.
- The now playing playlist overlay shows the matching list of beatmaps in the order they will be played.
- Song select has an actual random implementation (with associated user setting).

It seems like users will want to use the playlist overlay to potentially limit and order playback. Therefore, I'd propose that `MusicController` needs to maintain a list of playback order. In addition, one of the two paths forward:

- The game should randomise the order stored in `MusicController` if the user doesn't have a custom playlist selected, each startup.
- The game should use a stable method to decide the order, but we would add a "shuffle" button to the now playing overlay which advances in a non-sequential order (and this would be enabled by default).

I'm thinking the second path sounds best.

**Reply by frenzibyte** [reply]
I do agree with adding a shuffle button instead. In fact, there's already a discussion thread open for it (https://github.com/ppy/osu/discussions/13039).

**Reply by Jacie1krece** [reply]
I also agree with adding a shuffle button - that would mean that I won't need to re-start osu! to re-roll the `MusicController`, and the playlist function would be a great addition. But, I'm curious; why did `MusicController` go through the songs in "date added" order?

**Reply by peppy** [reply]
> But, I'm curious; why did `MusicController` go through the songs in "date added" order?

We changed the database backend, which has applied an arbitrary ordering to beatmaps. Previously it was random by chance.

**Reply by peppy** [reply]
Tracking via https://github.com/ppy/osu/issues/18169.
